cordapp-upgrades/README.md

```diff
@@ -0,0 +1,158 @@
+ `./gradlew issueBetweenNodes`, and old obligations can be settled using `./gradlew settleAllObligations`. The shell can
+ also be used to experiment.
+ 
+Note: If this example is compiled against Corda 4.0, then a bug will prevent this scenario from executing successfully.
+The issue is that the new version of `FinalityFlow` in 4.0 has a `targetPlatformVersion` gate that prevents the old API
+from being called if the app has set `targetPlatformVersion = 4`. To fix this, change the V2 workflows app to use
+`targetPlatformVersion = 3` instead. This issue is fixed in Corda 4.1.
+
+## Scenario 2: Contract Version Upgrade
+
+This section describes how to carry out an implicit contract upgrade. The upgraded contract in this case adds a new
+field to the obligation state (a defaulted flag), and a new command to govern transactions that set it. Additionally,
+version 4 of the workflows CorDapp provides a flow that can be used to set this flag.
+
+1. Deploy the nodes in their initial configuration by running `./gradlew deployNodes`. Start the network by
+ running `build/nodes/runnodes` and observe that all nodes can transact with each other (this can be done by running
+ `./gradlew issueBetweenNodes`, which will issue an obligation between each pair of nodes in the network).
+ 
+2. Shut down the network, then upgrade one of the nodes to the new version of the contract. This also requires
+ upgrading to the latest version of the workflows CorDapp, as older versions are compiled against the initial contract.
+ The upgrade can be carried out by running `scripts/upgradeNodes.sh workflows v3-can-default PartyA`
+ and `scripts/upgradeNodes.sh contracts v2-can-default PartyA`. Note: There is some additional configuration for the nodes running the new versions of the workflows CorDapp. This
+ configuration is used to control whether the new field in the contract state is set when the state is created. The no
+ downgrade rule will prevent nodes with only the old contract version from using the state if it has new fields set, and
+ so as a result the new field cannot be set until all nodes in the network are upgraded. To begin with, this configuration
+ prevents the field from being populated.
+ 
+3. Run the network (`build/nodes/runnodes`). In order to issue obligations between nodes, the contract attachment must be
+ present in the attachment store of each node (if this isn't the case, then an error will be hit when sending a transaction
+ with a version of the contract the node doesn't recognise). To put a copy of each contract version in the network in each
+ node's attachment store, run `./gradlew loadAttachments`. (This command opens all attachments on all nodes, and then
+ uploads each attachment to each node that doesn't already have it.)
+ 
+ 4. Check that obligations can be issued between nodes (again, `./gradlew issueBetweenNodes` can be used). It is
```

Comment:
4a Demonstrate a node (eg B) transacting with both V1 and V2 of the Contract
4b Demonstrate a node (eg C) transacting with V1 but failing fast with V2 of the Contract (until they explicitly install it on the node).

Reply:
I've updated the instructions to show that PartyA can issue an obligation naming PartyB (both on V2), and that PartyC can issue an obligation naming PartyA (C on V1, A on V2), as part of step 3. Step 4 then shows that if PartyB attempts to issue an obligation naming PartyC (B on V2, C on V1), an exception is raised on PartyC. I think that covers the scenarios you've mentioned here.